Added flag to enforce syscall dependencies

badnack opened the pull request:
**Issue**
Syzkaller often breaks dependencies across syscalls (expressed through the use of resources in Syzkaller programs). And because of this, Syzkaller struggles to build fuzzing inputs (i.e., programs) that would exercise deeper paths in the drivers. Syzkaller breaks syscalls dependencies because: 1) fuzzing mutation might remove random calls (and resources), 2) the resource generation itself is stochastic -- with a certain probability P, Syzkaller purposefully disregards syscall dependencies in an attempt to increase randomness, and 3) program minimization -- Syzkaller minimizes the programs that it generates, as they are initially quite big.

**Patch**
The patch addresses all these issues by adding a Boolean (called PromoteDep) in the validation module of Syzkaller. If the Boolean is set to true, certain measures are taken to enforce that IOCTL dependencies are respected in any generated program. In particular, Mutation does not break dependencies, Resource generation is no longer stochastic, and dependencies across syscalls are never disregarded, If a minimized program has broken dependencies, it gets discarded. The patch allows for different ways to enable the PromoteDep flag: 1) The configuration flag promote_syscalls_dependency. If enabled Syzkaller is instructed to not break dependencies, and 2) The flag dynamic_promote_syscalls_dependency. This flag contains a time expressed in minutes (e.g, 30). Once the manager starts it sets a timer with the value contained in this flag. Once the timer reaches the 0, the Boolean PromoteDep is switched (i.e., if it contained false it now contains true, and vice versa), and the timer starts again. This switch allows us to introduce more randomness in the generated programs. 

All flags are optional, and they don't have to be set.

dvyukov commented:
Hi Nilo,

How/when these tunables should be set? And when shouldn't they be set? Have you done any benchmarking? What are the results? Can this be enabled by default always?

We specifically tried to avoid lots of obscure knobs that nobody knows how to set right.

badnack commented:
Hello Dmitry,

Nice to meet you! Thanks for your message, let me answer to each one of your questions by topic:

Benchmarks
==
We did experiments during our fuzzing campaigns here at Qualcomm. In particular, we ran syzkaller with and without the proposed patch for a week-time on three different drivers. We observed that by enabling the option "dynamic_promote_syscalls_dependency" and setting it to 10 (minutes), syzkaller covered ~47% functions against  ~42% functions covered without the patch (this data was recorded by using GCOV), thus improving function coverage of, on average, ~5%. 

How/when to set the turnables
==
You can enable the flag "dynamic_promote_syscalls_dependency=10" and leave it always enabled, that's what we do in our fuzzing campaigns. What that does is: 1) Run syzkaller with the flag "PromoteDeps" set to true for 10 minutes, and then 2) Run syzkaller with the flag set to False for 10 minutes, and 3) repeat. In phase 1 Syzkaller generates fuzzing programs that do not contain any broken dependencies, thus triggering deeper states in the fuzzed driver.  During phase 2, Syzkaller behaves normally and it does not enforce any dependency. Phase 2 allows syzkaller to use the previous generated programs in phase 1 and apply mutation, etc..  By setting the "PromoteDeps" on and off we are able to generate a more complex and thorough fuzzing corpus. The knob could be set by default and never touched again. There are not really any cases in which the knob should not be set.

Please let me know of any other questions/feedback

dvyukov commented on syz-manager/manager.go:
```diff
@@ -863,9 +864,14 @@ func (mgr *Manager) saveRepro(res *manager.ReproResult) {
-		var err error
-		cprogText, err = repro.CProgram()
-		if err != nil {
+		cprog, err := csource.Write(repro.Prog, repro.Opts)
```
Please move this chunk into a separate PR, we could merge such small fixes/improvements much faster, and reduce diff of the main change.
The PR may include the MAINTAINERS change, and the newSignal logging change below.

a-nogikh replied:
This particular change actually looks like some merge conflict with recent changes in 1a7fb460b15fc14174513a2a2217c3a5b454ac49

dvyukov commented:
> Hello Dmitry,
> 
> Nice to meet you! Thanks for your message, let me answer to each one of your questions by topic:
> 
> ## Benchmarks
> We did experiments during our fuzzing campaigns here at Qualcomm. In particular, we ran syzkaller with and without the proposed patch for a week-time on three different drivers. We observed that by enabling the option "dynamic_promote_syscalls_dependency" and setting it to 10 (minutes), syzkaller covered ~47% functions against ~42% functions covered without the patch (this data was recorded by using GCOV), thus improving function coverage of, on average, ~5%.
> 
> ## How/when to set the turnables
> You can enable the flag "dynamic_promote_syscalls_dependency=10" and leave it always enabled, that's what we do in our fuzzing campaigns. What that does is: 1) Run syzkaller with the flag "PromoteDeps" set to true for 10 minutes, and then 2) Run syzkaller with the flag set to False for 10 minutes, and 3) repeat. In phase 1 Syzkaller generates fuzzing programs that do not contain any broken dependencies, thus triggering deeper states in the fuzzed driver. During phase 2, Syzkaller behaves normally and it does not enforce any dependency. Phase 2 allows syzkaller to use the previous generated programs in phase 1 and apply mutation, etc.. By setting the "PromoteDeps" on and off we are able to generate a more complex and thorough fuzzing corpus. The knob could be set by default and never touched again. There are not really any cases in which the knob should not be set.
> 
> All this said, I'd be happy to change/update the patch, if you think there is a better design to fix the issue of broken dependencies. Please let me know of any other questions/feedback.

Ok, this is good. Thanks.

1. Do we need to switch it globally? Can we just choose the mode locally for each program with some probability?
The global mutex-protected state looks very against design of the prog package. Choosing a mode randomly could simplify things and would be completely inline with the current design.
Since we also change Fuzzer.genFuzz, it may be a reasonable place to choose a mode and pass it to prog package.

2. I think we should try to enable to enable it by default w/o tunables if it's an improvement.

3. `mutateRate = 0.3` is pretty significant change (70% of generated programs instead of 5%). So I wonder if it accounts for the main improvement, and the rest is maybe less important.

badnack commented:
> > Hello Dmitry,
> > Nice to meet you! Thanks for your message, let me answer to each one of your questions by topic:
> > ## Benchmarks
> > We did experiments during our fuzzing campaigns here at Qualcomm. In particular, we ran syzkaller with and without the proposed patch for a week-time on three different drivers. We observed that by enabling the option "dynamic_promote_syscalls_dependency" and setting it to 10 (minutes), syzkaller covered ~47% functions against ~42% functions covered without the patch (this data was recorded by using GCOV), thus improving function coverage of, on average, ~5%.
> > ## How/when to set the turnables
> > You can enable the flag "dynamic_promote_syscalls_dependency=10" and leave it always enabled, that's what we do in our fuzzing campaigns. What that does is: 1) Run syzkaller with the flag "PromoteDeps" set to true for 10 minutes, and then 2) Run syzkaller with the flag set to False for 10 minutes, and 3) repeat. In phase 1 Syzkaller generates fuzzing programs that do not contain any broken dependencies, thus triggering deeper states in the fuzzed driver. During phase 2, Syzkaller behaves normally and it does not enforce any dependency. Phase 2 allows syzkaller to use the previous generated programs in phase 1 and apply mutation, etc.. By setting the "PromoteDeps" on and off we are able to generate a more complex and thorough fuzzing corpus. The knob could be set by default and never touched again. There are not really any cases in which the knob should not be set.
> > All this said, I'd be happy to change/update the patch, if you think there is a better design to fix the issue of broken dependencies. Please let me know of any other questions/feedback.
> 
> Ok, this is good. Thanks.
> 
> 1. Do we need to switch it globally? Can we just choose the mode locally for each program with some probability?
>    The global mutex-protected state looks very against design of the prog package. Choosing a mode randomly could simplify things and would be completely inline with the current design.
>    Since we also change Fuzzer.genFuzz, it may be a reasonable place to choose a mode and pass it to prog package.
> 2. I think we should try to enable to enable it by default w/o tunables if it's an improvement.
> 3. `mutateRate = 0.3` is pretty significant change (70% of generated programs instead of 5%). So I wonder if it accounts for the main improvement, and the rest is maybe less important.


1. The global switch is not strictly necessary, and we can achieve similar results by doing it locally using some probability as you mentioned. I can definitely do this change.
2. For sure, no problem.
3. I wouldn't say this is the main one, because even if we never mutate a program (for example), syscall dependencies are still very likely to be broken because of:  1) program minimization, 2) the fact that the resource generation is itself stochastic, and a resource might not be used even if it was previously created, and 3) there is a hard limit for the number of recursions allowed during the generation of a resource, which basically hinders the correct generation of programs whose syscalls dependencies are longer chains.

a-nogikh commented on prog/mutation.go:
```diff
@@ -207,9 +220,26 @@ func (ctx *mutator) removeCall() bool {
+		p.RemoveCall(idx)
+		success = true
+	}
+	return success
```
Does it ever return anything but `true`?

a-nogikh commented on prog/mutation.go:
```diff
@@ -246,13 +276,35 @@ func (ctx *mutator) mutateArg() bool {
+				idx--
+				if idx < 0 {
+					resizeGeneratedCalls(p, ctx.ncalls, c)
+					if len(p.Calls) > ctx.ncalls {
```
Nit: this check may better belong to `resizeGeneratedCalls` itself.

badnack replied:
you mean this block:
```
if len(p.Calls) > ctx.ncalls {
    panic(fmt.Sprintf("resizeGeneratedCalls failed: idx=%v calls=%v p.Calls=%v ncalls=%v",idx, len(calls), len(p.Calls), ctx.ncalls))
}
```

a-nogikh replied:
Yes

a-nogikh commented on prog/rand.go:
```diff
@@ -609,7 +627,9 @@ func (r *randGen) generateParticularCall(s *state, meta *Syscall) (calls []*Call
 		panic(fmt.Sprintf("generating no_generate call: %v", meta.Name))
 	}
 	c := MakeCall(meta, nil)
+	pushIoctlToStack(s, meta.Name)
```
It won't always be specifically ioctl, right? It can be just any syscall?

a-nogikh commented on prog/analysis.go:
```diff
@@ -12,19 +12,79 @@ import (
+	ioctlStack []string
+}
+
+func pushFieldToStack(s *state, fName string) {
```
It would really help to have some detailed comment regarding these new methods - what they are for / what kind of data `ioctlStack` actually represents.

badnack commented on pkg/fuzzer/job.go:
```diff
@@ -57,13 +57,17 @@ func mutateProgRequest(fuzzer *Fuzzer, rnd *rand.Rand) *queue.Request {
 		fuzzer.Config.NoMutateCalls,
 		fuzzer.Config.Corpus.Programs(),
 	)
+	if depsValidated && !newP.ValidateDeps() {
```
In my current patch, yes you are correct: A program with unbroken dependencies will never become a program with broken dependencies. However, I am working on a new patch that I am writing after collecting all the feedback I got, so this might change.

badnack commented on prog/expr.go:
```diff
@@ -72,7 +72,7 @@ func makeArgFinder(t *Target, c *Call, unionArg *UnionArg, parents parentStack)
 
 func (r *randGen) patchConditionalFields(c *Call, s *state) (extra []*Call, changed bool) {
-	if r.patchConditionalDepth > 1 {
+	if r.patchConditionalDepth > 1 && !IsPromoteDeps() {
```
You are right, I actually had notice that. Why is that the nested createResource calls are prohibited? In order to have a program with unbroken dependencies, nested createResource are necessary.

badnack commented on prog/generation.go:
```diff
@@ -26,9 +60,13 @@ func (target *Target) Generate(rs rand.Source, ncalls int, ct *ChoiceTable) *Pro
-	for len(p.Calls) > ncalls {
-		p.RemoveCall(ncalls - 1)
+	allowCalls := ncalls
+	if IsPromoteDeps() {
```
ok, I'll fix this.

badnack commented on prog/prog.go:
```diff
@@ -19,6 +19,18 @@ type Prog struct {
+	return p.validateDepsProtected()
+}
+
+func SetPromoteDeps(val bool) {
```
I am rewriting the patch using this approach :-)

badnack commented on prog/validation.go:
```diff
@@ -18,12 +21,39 @@ func init() {
+	toRet := true
+
+	if isPromoteDepsProtected() {
+		if err := p.validate(); err != nil {
```
Awesome! I'll update the patch, thank you!

badnack commented on prog/validation.go:
```diff
@@ -18,12 +21,39 @@ func init() {
 func (p *Prog) debugValidate() {
+	var stopIfDepBroken = isPromoteDepsProtected()
 	if debug {
+		setPromoteDepsProtected(false)
```
Now that I removed the global variable, these checks are gone altogether

badnack committed:
pkg/fuzzer, pkg/managerconfig, prog, syz-manager: logic to enforce syscall dependencies

Syzkaller often breaks dependencies across syscalls (e.g., due minimization, stochastic resource
generation, and mutation) when generating programs, thus failing to build fuzzing inputs that
exercise deep states in the target program.

This patch adds the logic to check whether syscall dependencies are broken in a given program.
Everytime a new program is to be generated, we set a flag called EnforceDeps to true with
a certain probability. If this flag is set, we enforce that the dependencies of each syscall
used in the program is respected.

badnack commented:
I applied all the requested changes

badnack commented:
Awesome, thank you!

a-nogikh commented:
Hi Nilo,

So:
20 runs of mainline syzkaller run for 24 hours (~28M execs each)
20 runs of syzkaller from this PR for 24 hours (~28M execs each)

I'd say that the syz-testbed results are very similar:

| Parameter            | Mainline | This PR |
|----------------------|----------|---------|
| crash types (median) | 133      | 136     |
| crash types (union of all runs) | 370   | 371  |
| crashes (median)     | 3130     | 3085    |
| coverage (median) | 306863   | 306842  |


Do you observe the improvements when you run the current code of the PR on the benchmarks you mentioned in https://github.com/google/syzkaller/pull/6131#issuecomment-3028620921?

badnack commented:
Hello @a-nogikh ,

Thanks for getting back to me. With the previous version of the patch, we ran fuzzing campaigns on a code base of ~75,000 functions and  we registered an improvement of ~3.7% function coverage (i.e., ~2,775 new functions). I am still running tests with the current version of the patch on the same codebase to compare results
